src/solvers/refinement/string_refinement.h

```diff
@@ -164,39 +160,61 @@ class string_refinementt: public bv_refinementt
-///   values
-/// \param last_concretized: initial value of the last concretized index
-template <typename T1, typename T2>
-void string_refinementt::pad_vector(
-  std::vector<T1> &concrete_array,
-  std::set<T2> &initialized,
-  T1 last_concretized) const
+/// \param initialized: set containing the indices of already concrete values
+template <typename T>
+void fill_in_vector(
+  std::vector<T> &concrete_array,
+  std::set<std::size_t> &initialized)
 {
   // Pad the concretized values to the left to assign the uninitialized
-  // values of result. The indices greater than concretize_limit are
-  // already assigned to last_concretized.
-  for(auto j=initialized.rbegin(); j!=initialized.rend();)
+  // values of result.
+  for(const auto &j : initialized)
   {
-    size_t i=*j;
-    // The leftmost index to pad is the value + 1 of the next element in
-    // 'initialized'. Since we cannot use the binary '+' operator on set
-    // iterators, we must increment the iterator here instead of in the
-    // for loop.
-    j++;
-    size_t leftmost_index_to_pad=(j!=initialized.rend()?*(j)+1:0);
-    // pad until we reach the next initialized index (right to left)
-    while(i>leftmost_index_to_pad)
-      concrete_array[(i--)-1]=last_concretized;
+    // Start concretizing from the left of `j` and pad from right to left until
+    // an initialized index (or 0) is reached
+    std::size_t i=j;
     INVARIANT(
-      i==leftmost_index_to_pad,
-      string_refinement_invariantt("Loop decrements i until it is not greater "
-        " than leftmost_index_to_pad"));
-    if(i>0)
-      last_concretized=concrete_array[i-1];
+      concrete_array.size()>j,
+      "set of initialized indices should not contain out of bound values");
+    while(i!=0 && initialized.find(i)==initialized.end())
+      concrete_array[--i]=concrete_array[j];
+  }
+}
+
+/// Utility function for concretization of strings. Copies concretized values to
+/// the left to initialize the unconcretized indices of concrete_array.
+/// \param initial_map: map containing the values of already concrete values
+/// \return a populated vector with value from initialized
+template <typename T>
+std::vector<T> fill_in_map_as_vector(
+  const std::map<std::size_t, T> &initial_map)
+{
+  if(initial_map.empty())
+    return std::vector<T>();
+
+  std::size_t last_index=initial_map.rbegin()->first;
+  std::vector<T> result(last_index+1);
+  // Pad the concretized values to the left to assign the uninitialized
+  // values of result.
+  for(auto pair=initial_map.rbegin(); pair!=initial_map.rend();)
+  {
+    const std::size_t i=pair->first;
+    const T value=pair->second;
+    // We must increment the iterator here instead of in the for loop so that
+    // we can get the leftmost index to pad
+    pair++;
+    const std::size_t leftmost_index_to_pad=
+      pair!=initial_map.rend()?pair->first:0;
+    // pad down to the leftmost index to pad
+    for(std::size_t j=i; j+1!=leftmost_index_to_pad; j--)
```

Comment:
If "down to" is inclusive then it should be correct as the last  where the inequality holds is `leftmost_index_to_pad`